dolphin/drivers/huawei/oceanstor/rest_client.py

```diff
@@ -0,0 +1,251 @@
+            LOG.debug('Replace URL: \n'
+                      'Old URL: %(old_url)s\n,'
+                      'New URL: %(new_url)s\n.',
+                      {'old_url': old_url,
+                       'new_url': self.url})
+            result = self.do_call(url, data, method,
+                                  log_filter_flag=log_filter_flag)
+            if result['error']['code'] in consts.RELOGIN_ERROR_PASS:
+                result['error']['code'] = 0
+        return result
+
+    def logout(self):
+        """Logout the session."""
+        url = "/sessions"
+        if self.url:
+            result = self.do_call(url, None, "DELETE")
+            self._assert_rest_result(result, _('Logout session error.'))
+
+    def _assert_rest_result(self, result, err_str):
+        if result['error']['code'] != 0:
+            msg = (_('%(err)s\nresult: %(res)s.') % {'err': err_str,
+                                                     'res': result})
+            LOG.error(msg)
+            raise exception.StorageBackendException(reason=msg)
+
+    def _assert_data_in_result(self, result, msg):
+        if 'data' not in result:
+            err_msg = _('%s "data" is not in result.') % msg
+            LOG.error(err_msg)
+            raise exception.StorageBackendException(reason=err_msg)
+
+    def get_storage(self):
+        url = "/system/"
+        result = self.call(url, method='GET')
+
+        msg = _('Create lun error.')
+        self._assert_rest_result(result, msg)
+        self._assert_data_in_result(result, msg)
+
+        return result['data']
+
+    def get_all_volumes(self):
+        url = "/lun"
+        result = self.call(url, None, "GET", log_filter_flag=True)
+        msg = _('Query resource volume error.')
+        self._assert_rest_result(result, msg)
+        self._assert_data_in_result(result, msg)
+        return result['data']
+
+    def get_all_pools(self):
+        url = "/storagepool"
+        result = self.call(url, None, "GET", log_filter_flag=True)
+        msg = _('Query resource pool error.')
+        self._assert_rest_result(result, msg)
+        self._assert_data_in_result(result, msg)
+        return result['data']
+
+    def get_pool_info(self, pool_name=None, pools=None):
+        info = {}
+        if not pool_name:
+            return info
+
+        for pool in pools:
+            if pool_name.strip() != pool['NAME']:
+                continue
+
+            if pool.get('USAGETYPE') == consts.FILE_SYSTEM_POOL_TYPE:
+                break
+
+            info['ID'] = pool['ID']
+            info['CAPACITY'] = pool.get('DATASPACE', pool['USERFREECAPACITY'])
+            info['TOTALCAPACITY'] = pool.get('USERTOTALCAPACITY', '0')
+            info['TIER0CAPACITY'] = pool.get('TIER0CAPACITY', '0')
+            info['TIER1CAPACITY'] = pool.get('TIER1CAPACITY', '0')
+            info['TIER2CAPACITY'] = pool.get('TIER2CAPACITY', '0')
+
+        return info
+
+    def get_pool_id(self, pool_name):
```

Comment:
This seems to be complicated. Is this done because there is no API for getting pool details from pool Name...
If there was pool name, then there was a discovery of pool..Can you please explain this..This function is in-efficient and maybe because of vendor APIs but plz explain

Reply:
Refactored to optimize

Reply:
OK